app/uk/gov/hmrc/cdsreimbursementclaim/services/ClaimTransformerService.scala

```diff
@@ -609,53 +603,45 @@ object DefaultClaimTransformerService {
-            MrnDetail(
-              MRNNumber = Some(mrn.value),
-              acceptanceDate = Some(acceptanceDate),
-              declarantReferenceNumber = displayDeclaration.displayResponseDetail.declarantReferenceNumber,
-              mainDeclarationReference = Some(true),
-              procedureCode = Some(displayDeclaration.displayResponseDetail.procedureCode),
-              declarantDetails = Some(declarationDetails),
-              accountDetails = transformToMaybeAccountDetail(displayDeclaration.displayResponseDetail.accountDetails),
-              consigneeDetails = Some(consigneeDetails),
-              bankDetails = Some(bankDetails),
-              NDRCDetails = Some(ndrcDetails)
-            )
-          }
+
+    val details = completeClaim.displayDeclaration.toList.flatMap(displayDeclaration =>
+      multipleClaimsAnswer(completeClaim).map { case (mrn, claim) =>
+        (
+          setAcceptanceDate(displayDeclaration.displayResponseDetail.acceptanceDate),
+          setDeclarantDetails(displayDeclaration.displayResponseDetail.declarantDetails),
+          setConsigneeDetails(displayDeclaration.displayResponseDetail.consigneeDetails),
+          buildBankDetails(displayDeclaration.displayResponseDetail.bankDetails, completeClaim),
+          makeNdrcDetails(claim)
+        ).mapN { case (acceptanceDate, declarationDetails, consigneeDetails, bankDetails, ndrcDetails) =>
+          MrnDetail(
+            MRNNumber = Some(mrn.value),
+            acceptanceDate = Some(acceptanceDate),
+            declarantReferenceNumber = displayDeclaration.displayResponseDetail.declarantReferenceNumber,
+            mainDeclarationReference = Some(true),
+            procedureCode = Some(displayDeclaration.displayResponseDetail.procedureCode),
+            declarantDetails = Some(declarationDetails),
+            accountDetails = transformToMaybeAccountDetail(displayDeclaration.displayResponseDetail.accountDetails),
+            consigneeDetails = Some(consigneeDetails),
+            bankDetails = Some(bankDetails),
+            NDRCDetails = Some(ndrcDetails)
+          )
         }
-      case None                     => List()
-    }
-    if (details.isEmpty) {
-      invalidNel("could not obtain any MRN details")
-    } else {
-      val (valid, invalid) = details.partition(_.isValid)
-      if (!invalid.isEmpty) {
-        val errors = invalid.collect { case e: Invalid[NonEmptyList[String]] => e }
-        invalidNel(
-          s"there is at least one claim which has failed validation: ${errors.map(s => s.e.toList).mkString("|")}"
-        )
-      } else {
-        val mrnDetail: List[MrnDetail] = valid.collect { case Valid(value) => value }
-        Valid(mrnDetail)
       }
-    }
+    )
+
+    details
+      .ensuring(_.nonEmpty)
+      .sequence
+      .leftMap(errors =>
+        NonEmptyList.one(s"there is at least one claim which has failed validation: ${errors.toList.mkString("|")}")
+      )
   }
```

Comment:
@steve-edmans I replaced all this by `sequence`.
All you needed to is `details.sequence`.
This is common way in Scala how to from List(F[A1], F[A2]) get something as F(List(A1, A2))